Mathlib/Algebra/Lie/Semisimple.lean

```diff
@@ -120,31 +111,470 @@ theorem hasTrivialRadical_iff_no_abelian_ideals :
+    apply_fun Subtype.val at this
+    dsimp at this
+    rwa [this, zero_add]
+  -- To show that `y` is in the center of `J`,
+  -- we show that any `j ∈ J` brackets to `0` with `z` and with `x = y + z`.
+  -- By a simple computation, that implies `⁅j, y⁆ = 0`, for all `j`, as desired.
+  intro j
+  suffices ⁅(j : L), z⁆ = 0 ∧ ⁅(j : L), y + z⁆ = 0 by
+    rw [lie_add, this.1, add_zero] at this
+    ext
+    exact this.2
+  rw [← LieSubmodule.mem_bot (R := R) (L := L), ← LieSubmodule.mem_bot (R := R) (L := L)]
+  constructor
+  -- `j` brackets to `0` with `z`, since `⁅j, z⁆` is contained in `⁅J, K⁆ ≤ J ⊓ K`,
+  -- and `J ⊓ K = ⊥` by the independence of the atoms.
+  · apply (hS.setIndependent.disjoint_sSup (hs hJs) hs'S (Finset.not_mem_erase _ _)).le_bot
+    apply LieSubmodule.lie_le_inf
+    apply LieSubmodule.lie_mem_lie _ _ j.2
+    simpa only [K, Finset.sup_id_eq_sSup] using hz
+  -- By similar reasoning, `j` brackets to `0` with `x = y + z ∈ I`, if we show `J ⊓ I = ⊥`.
+  suffices J ⊓ I = ⊥ by
+    apply this.le
+    apply LieSubmodule.lie_le_inf
+    exact LieSubmodule.lie_mem_lie _ _ j.2 hx
+  -- Indeed `J ⊓ I = ⊥`, since `J` is an atom that is not contained in `I`.
+  apply ((hS.isAtom J (hs hJs)).le_iff.mp _).resolve_right
+  · contrapose! hJI
+    rw [← hJI]
+    exact inf_le_right
+  exact inf_le_left
+termination_by s => s.card
+decreasing_by
+  simp_wf
+  exact Finset.card_lt_card hs'
+
+/--
+In a semisimple Lie algebra,
+Lie ideals that are contained in the supremum of a collection of atoms
+are themselves the supremum of a subcollection of those atoms.
+
+The proof is by a compactness argument, reducing to finite collections of atoms.
+In the finite case, the proof is given in `atomistic_of_finset`.
+-/
+lemma atomistic (J : LieIdeal R L) (hJ : J ≤ sSup S) : ∃ T ⊆ S, J = sSup T := by
+  obtain ⟨C, hC, rfl⟩ := IsCompactlyGenerated.exists_sSup_eq J
+  suffices ∀ J' : LieIdeal R L, IsCompactElement J' → J' ≤ sSup S → ∃ T ⊆ S, J' = sSup T by
+    choose T hT₁ hT₂ using this
+    use sSup {T J' h₁ h₂ | (J' ∈ C) (h₁ : IsCompactElement J') (h₂ : J' ≤ sSup S)}
+    constructor
+    · apply _root_.sSup_le
+      rintro _ ⟨J', -, h₁, h₂, rfl⟩
+      apply hT₁
+    · apply le_antisymm
+      · apply _root_.sSup_le
+        intro J' hJ'
+        rw [hT₂ J' (hC _ hJ') ((le_sSup hJ').trans hJ)]
+        apply sSup_le_sSup
+        apply _root_.le_sSup
+        use J', hJ', hC _ hJ', (le_sSup hJ').trans hJ
+      · simp only [Set.sSup_eq_sUnion, sSup_le_iff, Set.mem_sUnion, Set.mem_setOf_eq,
+          forall_exists_index, and_imp]
+        rintro I T J' hJ'C hJ' hJ'S rfl hIT
+        apply (le_sSup hIT).trans
+        rw [← hT₂]
+        exact le_sSup hJ'C
+  clear hJ hC C
+  intro J hJ hJS
+  obtain ⟨s, hs₁, hs₂⟩ := hJ S hJS
+  obtain ⟨t, ht, rfl⟩ := hS.atomistic_of_finset s hs₁ J hs₂
+  refine ⟨t, ?_, Finset.sup_id_eq_sSup t⟩
+  refine Set.Subset.trans ?_ hs₁
+  simpa only [Finset.coe_subset] using ht
+
```

Comment:
I think we can get more help from the lattice theory library here. For example:
```lean

lemma isAtomistic_of_sSup_eq_top (h : sSup S = ⊤) : IsAtomistic (LieIdeal R L) := by
  refine ⟨fun J ↦ ?_⟩
  obtain ⟨s, hs, hs'⟩ := atomistic hS J (by aesop)
  exact ⟨s, hs', fun I hI ↦ hS.isAtom I (hs hI)⟩

lemma complementedLattice_of_sSup_eq_top (h : sSup S = ⊤) : ComplementedLattice (LieIdeal R L) := by
  have := isAtomistic_of_sSup_eq_top hS h
  apply complementedLattice_of_isAtomistic

```
and then I presume some of the below lemmas should also become easy / redundant.

Reply:
Incidentally it's not clear to me why `complementedLattice_of_isAtomistic` isn't an instance.

Reply:
Nice!

I guess it is not an instance as a remnant of the Lean 3 days, because the other direction is already an instance, and it would create a loop.

We could try to make it an instance in another PR...

Reply:
> and then I presume some of the below lemmas should also become easy / redundant.

Actually, those lemmas are all working towards showing that the lattice is distributive. I've now indicated that in a comment.

Reply:
I now completely abstracted these lemmas into lattice theory.